File: src/game/adventure/main/GameLogic.java
package game.adventure.main;

import game.adventure.gameobjects.Enemy;
import game.adventure.gameobjects.Item;
import game.adventure.objects.ItemPair;
import game.adventure.util.Constants;
import game.adventure.util.Direction;
import game.adventure.util.HelperFunctions;

import java.util.*;

public class GameLogic {
    private static final Scanner s = new Scanner(System.in);

    private GameLogic() { }

    public static void start() {
        AdventureManager.getCurrentAdventure().init();
        takeInput();
    }

    private static void takeInput() {
        String input;
        System.out.println("Press h for help!");

        String n = HelperFunctions.startsWithVowel(AdventureManager.getCurrentRoom().getName()) ? "n " : " ";

        System.out.println("You find yourself in a" + n + AdventureManager.getCurrentRoom().getName());
        do {
            AdventureManager.getCurrentRoom().setHasVisited(true);
            System.out.print("> ");
            input = s.nextLine().toLowerCase();
            ParseCommand(input);
        } while (!input.equals("q") && !input.equals("quit"));
    }

    private static void ParseCommand(String input) {
        List<String> wordList;
        wordList = wordList(input);

        try {
            wordList.get(0);
        } catch (IndexOutOfBoundsException e) {
            System.out.println("You must enter a command!");
            return;
        }

        switch (wordList.get(0)) {
            case "n":
            case "north":
            case "s":
            case "south":
            case "e":
            case "east":
            case "w":
            case "west":
                getDirection(wordList);
                break;
            case "i":
            case "inv":
            case "inventory":
                inventoryCommand();
                break;
            case "h":
            case "help":
                helpCommand();
                break;
            case "l":
            case "look":
                lookCommand(wordList);
                break;
            case "t":
            case "take":
                takeCommand(wordList);
                break;
            case "d":
            case "drop":
                dropCommand(wordList);
                break;
            case "u":
            case "use":
                useCommand(wordList);
                break;
            case "eq":
            case "equip":
                equipCommand(wordList);
                break;
            case "q":
            case "quit":
                break;
            default:
                System.out.println("I don't understand " + wordList.get(0));
                break;
        }
    }

    static void playerDeath(String killer) {
        System.out.println("The " + killer + " killed you!");
        System.out.println("rip bozo");
        System.exit(0);
    }

    private static void equipCommand(List<String> wordList) {
        if (wordList.size() == 1) {
            AdventureManager.getPlayer().removeEquip();
            System.out.println("Removed your equip");
            return;
        }

        String o = HelperFunctions.parseItemName(wordList);

        if (!AdventureManager.getPlayer().hasItem(o)) {
            System.out.println("You don't have the item " + o + "!");
            return;
        }

        if (!AdventureManager.getPlayer().getItem(o).isEquippable()) {
            System.out.println("You can't equip " + o + "!");
            return;
        }

        AdventureManager.getPlayer().setEquip(o);
        System.out.println("Equipped " + o);
    }

    private static void inspectItem(List<String> wordList) {
        String o = HelperFunctions.parseItemName(wordList);

        for (Item i : AdventureManager.getCurrentRoom().getItems()) {
            if (i.getName().equalsIgnoreCase(o)) {
                System.out.println(i.getName() + ": " + i.getDescription());
                return;
            }
        }
        for (Item i : AdventureManager.getPlayer().getInventory()) {
            if (i.getName().equalsIgnoreCase(o)) {
                System.out.println(i.getName() + ": " + i.getDescription());
                return;
            }
        }
        for (Enemy e : AdventureManager.getCurrentRoom().getEnemies()) {
            if (e.getName().equalsIgnoreCase(o)) {
                System.out.println(e.getName() + " (" + e.getHealth() + " hp " + e.getDamage() + " damage): " + e.getDescription());
                return;
            }
        }
        System.out.println("I don't know what " + o + " is!");
    }

    private static void helpCommand() {
        System.out.println("""
                Quit - "q"
                Look - "l" or "l <item>"
                Inventory - "i"
                Take - "t <item>"
                Drop - "d <item>"
                Equip - "eq <item>"
                Use - "u <item1> on <item2>"
                Movement - "n", "s", "e", and "w\"""");
    }

    private static void lookItemsCommand() {
        if (AdventureManager.getCurrentRoom().getItems().isEmpty()) {
            System.out.println("There are no items in this room!");
            return;
        }

        System.out.print("Items near you: ");
        for (int i = 0; i < AdventureManager.getCurrentRoom().getItems().size(); i++) {
            String separator = i == AdventureManager.getCurrentRoom().getItems().size() - 1 ? "." : ", ";

            System.out.print(AdventureManager.getCurrentRoom().getItem(i).getName() + separator);
        }
        System.out.println();

    }

    private static void lookCommand(List<String> wordList) {
        if (wordList.size() > 1) {
            inspectItem(wordList);
            return;
        }

        System.out.println(AdventureManager.getCurrentRoom().getName() + ": " + AdventureManager.getCurrentRoom().getDescription());

        lookItemsCommand();

        HelperFunctions.lookCardinal();

        if (AdventureManager.getCurrentRoom().getEnemies().isEmpty()) return;

        for (int i = 0; i < AdventureManager.getCurrentRoom().getEnemies().size(); i++) {
            Enemy e = AdventureManager.getCurrentRoom().getEnemies().get(i);
            System.out.println("Enemies near you: ");
            System.out.println(e.getName() + ": " + e.getDescription());
        }
    }

    private static void getInventory() {
        if (AdventureManager.getPlayer().getInventory().isEmpty() || (AdventureManager.getPlayer().getInventory().size() == 1 && !AdventureManager.getPlayer().getEquip().equals(Constants.GENERIC_ITEM))) {
            System.out.println("You have nothing in your inventory!");
            return;
        }

        System.out.println("Your inventory: ");
        for (int i = 0; i < AdventureManager.getPlayer().getInventory().size(); i++) {
            String nameSuffix = AdventureManager.getPlayer().getItem(i).getDamage() == -1 ? ": " : " (" + AdventureManager.getPlayer().getItem(i).getDamage() + " damage): ";

            System.out.println(AdventureManager.getPlayer().getItem(i).getName() + nameSuffix + AdventureManager.getPlayer().getItem(i).getDescription());
        }
    }

    private static void inventoryCommand() {
        System.out.println("Health: " + AdventureManager.getPlayer().getHealth());
        int damage = AdventureManager.getPlayer().getDamage();
        System.out.println("Damage: " + damage);
        System.out.println();

        getInventory();

        if (AdventureManager.getPlayer().hasEquip()) {
            System.out.println("\nEquipped item: \n" + AdventureManager.getPlayer().getEquip().getName() + " (" + AdventureManager.getPlayer().getDamage() + " damage): " + AdventureManager.getPlayer().getEquip().getDescription());
        }
    }

    private static List<String> wordList(String input) {
        String delims = "\s\t,.:;?!\"'";
        List<String> strList = new ArrayList<>();
        StringTokenizer tokenizer = new StringTokenizer(input, delims);
        String t;

        while (tokenizer.hasMoreTokens()) {
            t = tokenizer.nextToken();
            strList.add(t);
        }

        return strList;
    }

    private static void useCommand(List<String> wordList) {
        String o, oo;

        if (wordList.size() < 4) {
            System.out.println("Not enough parameters!");
            return;
        }

        try {
            StringBuilder builder = new StringBuilder();
            for (int i = 1; i < wordList.size(); i++) {
                if (wordList.get(i).equals("on")) {
                    wordList.remove(i);
                    break;
                }

                builder.append(wordList.get(i));
                builder.append(" ");
                wordList.remove(i);
                i--;
            }

            o = builder.toString().trim();

            oo = HelperFunctions.parseItemName(wordList);

            Item item1 = HelperFunctions.getItemFromName(o);
            Item item2 = HelperFunctions.getItemFromName(oo);

            if (!((AdventureManager.getPlayer().hasItem(o) || (AdventureManager.getCurrentRoom().hasItem(o) && !item1.canPickUp())) &&
                    ((AdventureManager.getPlayer().hasItem(oo) || (AdventureManager.getCurrentRoom().hasItem(oo) && !item2.canPickUp()))))) {
                System.out.println("You don't have that item!");
                return;
            }

            AdventureManager.getCurrentAdventure().useFetcher().get(new ItemPair(item1, item2)).run();
        } catch (IndexOutOfBoundsException e) {
            System.out.println("Not enough information provided!");
        }
    }

    private static void takeCommand(List<String> wordList) {
        if (wordList.size() <= 1) {
            System.out.println("Please specify the item to take!");
            return;
        }

        String o;

        o = HelperFunctions.parseItemName(wordList);

        if (!AdventureManager.getCurrentRoom().hasItem(o)) {
            System.out.println("Can't find that item!");
            return;
        }

        if (!AdventureManager.getCurrentRoom().getItem(o).canPickUp()) {
            System.out.println("You can't take that item!");
            return;
        }

        AdventureManager.getPlayer().addItem(AdventureManager.getCurrentRoom().getItem(o));
        AdventureManager.getCurrentRoom().removeItem(HelperFunctions.getItemFromName(o));
        System.out.println("You took the " + AdventureManager.getPlayer().getItem(o).getName() + ".");

    }

    private static void dropCommand(List<String> wordList) {
        if (wordList.size() <= 1) {
            System.out.println("What item would you like to drop?");
            return;
        }

        String o;

        o = HelperFunctions.parseItemName(wordList);

        if (!AdventureManager.getPlayer().hasItem(o)) {
            System.out.println("You don't have that item!");
        }

        if (AdventureManager.getPlayer().isEquip(HelperFunctions.getItemFromName(o))) {
            AdventureManager.getPlayer().removeEquip();
        }

        AdventureManager.getCurrentRoom().addItem(AdventureManager.getPlayer().getItem(o));
        AdventureManager.getPlayer().removeItem(o);
        System.out.println("You dropped the " + AdventureManager.getCurrentRoom().getItem(o).getName() + ".");

    }

    private static void getDirection(List<String> input) {
        boolean roomChanged = false;
        String intendedDir = input.get(0);
        String n;

        for (Direction direction : Direction.values()) {
            if (!intendedDir.equalsIgnoreCase(direction.getString()) && !intendedDir.equalsIgnoreCase(direction.getFirst())) continue;
            if (AdventureManager.getCurrentRoom().getDirection(direction).isEmpty()) {
                System.out.println("You can't go " + direction.getString() + " here!");
                return;
            }
            roomChanged = true;
            AdventureManager.setCurrentRoom(AdventureManager.getCurrentRoom().getDirection(direction));
            n = HelperFunctions.startsWithVowel(AdventureManager.getCurrentRoom().getName()) ? "n " : " ";
            System.out.println("You find yourself in a" + n + AdventureManager.getCurrentRoom().getName());
        }
        if (!AdventureManager.getCurrentRoom().getEnemies().isEmpty() && roomChanged) {
            new Combat();
        }
    }
    
}

File: src/game/adventure/main/GameLogic.kt
package game.adventure.main

import game.adventure.gameobjects.Enemy
import game.adventure.main.HelperFunctions.getMap
import game.adventure.main.HelperFunctions.lookCardinal
import game.adventure.main.HelperFunctions.parseItemName
import game.adventure.main.HelperFunctions.startsWithVowel
import game.adventure.utils.Constants
import game.adventure.utils.GameInit
import game.adventure.utils.Utils.useFetcher
import java.util.*
import kotlin.system.exitProcess

class GameLogic {
    fun takeInput() {
        var input: String
        println("Press h for help!")
        val n = if (startsWithVowel(currentRoom.name)) "n " else " "
        println("You find yourself in a" + n + currentRoom.name)
        do {
            currentRoom.hasVisited = true
            print("> ")
            input = s.nextLine().lowercase(Locale.getDefault())
            ParseCommand(input)
        } while (input != "q" && input != "quit")
    }

    fun ParseCommand(input: String?) {
        val wordList: MutableList<String?>
        wordList = wordList(input)
        try {
            wordList[0]
        } catch (e: IndexOutOfBoundsException) {
            println("You must enter a command!")
            return
        }
        try {
            when (wordList[0]) {
                "n", "north", "s", "south", "e", "east", "w", "west" -> getDirection(wordList)
                "i", "inv", "inventory" -> inventoryCommand()
                "h", "help" -> helpCommand()
                "l", "look" -> lookCommand(wordList)
                "t", "take" -> takeCommand(wordList)
                "d", "drop" -> dropCommand(wordList)
                "u", "use" -> useCommand(wordList)
                "eq", "equip" -> equipCommand(wordList)
                "a", "attack" -> attackCommand()
                "q", "quit" -> {}
                else -> println("I don't understand " + wordList[0])
            }
        } finally {
            if (wordList.size != 0) {
                when (wordList[0]) {
                    "l", "look", "n", "north", "s", "south", "e", "east", "w", "west" -> {}
                    else -> enemyAttack()
                }
            }
        }
    }

    private fun playerDeath(killer: String) {
        println("The $killer killed you!")
        println("rip bozo")
        exitProcess(0)
    }

    private fun enemyAttack() {
        for (e in currentRoom.getEnemies()) {
            Constants.player.health = Constants.player.health - e.damage
            if (Constants.player.health <= 0) {
                playerDeath(e.name)
                return
            }
            println("The " + e.name + " did " + e.damage + " damage to you!")
        }
    }

    private fun attackCommand() {
        val toRemove: MutableList<Enemy> = ArrayList()
        if (currentRoom.getEnemies().isEmpty()) {
            println("There are no enemies in this room!")
            return
        }
        for (e in currentRoom.getEnemies()) {
            val damage = if (Constants.player.getEquip() == -1) 1 else Constants.player.damage
            e.health = e.health - damage
            println("You did " + damage + " damage to the " + e.name)
            if (e.health <= 0) {
                println("You killed the " + e.name)
                toRemove.add(e)
                print("It dropped: ")
                for (i in e.getDrops().indices) {
                    val drop = e.getDrops()[i]
                    currentRoom.addItem(drop)
                    val separator = if (i == e.getDrops().size - 1) "." else ", "
                    println(e.getDrops()[i].name + separator)
                }
            }
        }
        try {
            currentRoom.getEnemies().removeAll(toRemove)
        } catch (ignore: NullPointerException) {
        }
    }

    private fun equipCommand(wordList: List<String?>) {
        if (wordList.size == 1) {
            Constants.player.removeEquip()
            println("Removed your equip")
            return
        }
        val o = parseItemName(wordList)
        if (!Constants.player.hasItem(o)) {
            println("You don't have the item $o!")
            return
        }
        Constants.player.setEquip(o)
        println("Equipped $o")
    }

    private fun inspectItem(wordList: List<String?>?) {
        val o = parseItemName(wordList!!)
        for (i in currentRoom.getItems()) {
            if (i.name.equals(o, ignoreCase = true)) {
                println(i.name + ": " + i.description)
                return
            }
        }
        for (i in Constants.player.inventory) {
            if (i.name.equals(o, ignoreCase = true)) {
                println(i.name + ": " + i.description)
                return
            }
        }
        for (e in currentRoom.getEnemies()) {
            if (e.name.equals(o, ignoreCase = true)) {
                println(e.name + " (" + e.health + " hp " + e.damage + " damage): " + e.description)
                return
            }
        }
        println("I don't know what $o is!")
    }

    private fun helpCommand() {
        println(/* !!! Hit visitElement for element type: class org.jetbrains.kotlin.nj2k.tree.JKErrorExpression !!! */)
    }

    private fun lookItemsCommand() {
        if (currentRoom.getItems().isEmpty()) {
            println("There are no items in this room!")
            return
        }
        print("Items near you: ")
        for (i in currentRoom.getItems().indices) {
            val separator = if (i == currentRoom.getItems().size - 1) "." else ", "
            print(currentRoom.getItem(i).name + separator)
        }
        println()
    }

    private fun lookCommand(wordList: List<String?>) {
        if (wordList.size > 1) {
            inspectItem(wordList)
            return
        }
        println(currentRoom.name + ": " + currentRoom.description)
        lookItemsCommand()
        lookCardinal()
        if (currentRoom.getEnemies().isEmpty()) return
        for (i in currentRoom.getEnemies().indices) {
            val e = currentRoom.getEnemies()[i]
            println("Enemies near you: ")
            println(e.name + ": " + e.description)
        }
    }

    private val inventory: Unit
        get() {
            if (Constants.player.inventory.isEmpty() || Constants.player.inventory.size == 1 && Constants.player.getEquip() != -1) {
                println("You have nothing in your inventory!")
                return
            }
            println("Your inventory: ")
            for (i in Constants.player.inventory.indices) {
                if (i == Constants.player.getEquip()) {
                    continue
                }
                val nameSuffix =
                    if (Constants.player.getItem(i).damage == -1) ": " else " (" + Constants.player.getItem(i).damage + " damage): "
                println(Constants.player.getItem(i).name + nameSuffix + Constants.player.getItem(i).description)
            }
        }

    private fun inventoryCommand() {
        println("Health: " + Constants.player.health)
        val damage = if (Constants.player.getEquip() == -1) 1 else Constants.player.damage
        println("Damage: $damage")
        println()
        inventory
        if (Constants.player.getEquip() != -1) {
            println(
                """
Equipped item: 
${Constants.player.equipItem.name} (${Constants.player.damage} damage): ${Constants.player.equipItem.description}"""
            )
        }
    }

    private fun wordList(input: String?): MutableList<String?> {
        val delims = " \t,.:;?!\"'"
        val strList: MutableList<String?> = ArrayList()
        val tokenizer = StringTokenizer(input, delims)
        var t: String?
        while (tokenizer.hasMoreTokens()) {
            t = tokenizer.nextToken()
            strList.add(t)
        }
        return strList
    }

    private fun useCommand(wordList: MutableList<String?>) {
        val o: String
        val oo: String
        try {
            val builder = StringBuilder()
            var i = 1
            while (i < wordList.size) {
                if (wordList[i] == "on") {
                    wordList.removeAt(i)
                    break
                }
                builder.append(wordList[i])
                builder.append(" ")
                wordList.removeAt(i)
                i--
                i++
            }
            o = builder.toString().trim { it <= ' ' }
            oo = parseItemName(wordList)
            if (!((Constants.player.hasItem(o) || currentRoom.hasItem(o)) && (Constants.player.hasItem(oo) || currentRoom.hasItem(
                    oo
                )))
            ) {
                println("You don't have that item!")
                return
            }
            useFetcher(o, oo).run()
        } catch (e: IndexOutOfBoundsException) {
            println("Not enough information provided!")
        }
    }

    private fun takeCommand(wordList: List<String?>) {
        if (wordList.size <= 1) {
            println("What item would you like to take?")
            return
        }
        val o: String = parseItemName(wordList)
        if (!currentRoom.hasItem(o)) {
            println("Can't find that item!")
            return
        }
        if (!currentRoom.getItem(o).canPickUp()) {
            println("You can't take that item!")
            return
        }
        Constants.player.addItem(currentRoom.getItem(o))
        currentRoom.removeItem(o)
        println("You took the " + Constants.player.getItem(o).name + ".")
    }

    private fun dropCommand(wordList: List<String?>) {
        if (wordList.size <= 1) {
            println("What item would you like to drop?")
            return
        }
        val o: String = parseItemName(wordList)
        if (!Constants.player.hasItem(o)) {
            println("You don't have that item!")
        }
        if (Constants.player.isEquip(o)) {
            Constants.player.setEquip(-1)
        }
        currentRoom.addItem(Constants.player.getItem(o))
        Constants.player.removeItem(o)
        println("You dropped the " + currentRoom.getItem(o).name + ".")
    }

    private fun getDirection(input: List<String?>) {
        var roomChanged = false
        val direction = input[0]
        val n: String
        try {
            when (direction) {
                "n", "north" -> {
                    if (currentRoom.n == -1) {
                        println("You can't go north here!")
                        return
                    }
                    roomChanged = true
                    currentRoom = getMap(currentRoom.n)
                    n = if (startsWithVowel(currentRoom.name)) "n " else " "
                    println("You find yourself in a" + n + currentRoom.name)
                }

                "s", "south" -> {
                    if (currentRoom.s == -1) {
                        println("You can't go south here!")
                        return
                    }
                    roomChanged = true
                    currentRoom = getMap(currentRoom.s)
                    n = if (startsWithVowel(currentRoom.name)) "n " else " "
                    println("You find yourself in a" + n + currentRoom.name)
                }

                "e", "east" -> {
                    if (currentRoom.e == -1) {
                        println("You can't go east here!")
                        return
                    }
                    roomChanged = true
                    currentRoom = getMap(currentRoom.e)
                    n = if (startsWithVowel(currentRoom.name)) "n " else " "
                    println("You find yourself in a" + n + currentRoom.name)
                }

                "w", "west" -> {
                    if (currentRoom.w == -1) {
                        println("You can't go west here!")
                        return
                    }
                    roomChanged = true
                    currentRoom = getMap(currentRoom.w)
                    n = if (startsWithVowel(currentRoom.name)) "n " else " "
                    println("You find yourself in a" + n + currentRoom.name)
                }
            }
        } finally {
            if (currentRoom.getEnemies().isEmpty() || !roomChanged) return
            for (i in currentRoom.getEnemies().indices) {
                val e = currentRoom.getEnemies()[i]
                print("Enemies near you: ")
                val separator = if (i == currentRoom.getEnemies().size - 1) "." else ", "
                println(e.name + " (" + e.health + " hp " + e.damage + " damage)" + separator)
            }
        }
    }

    companion object {
        private val s = Scanner(System.`in`)
        val gi = GameInit()
        var currentRoom = getMap(0)
    }
}